File: package/main/src/URL/parseQueryString.ts
/**
 * Parses a query string into a key-value record.
 *
 * Accepts either a full URL or a raw query string
 * (with or without leading "?").
 *
 * @param query - The query string or URL to parse
 * @returns A record of key-value pairs from the query string
 *
 * @remarks
 * **Prototype pollution warning:** This function does not filter out
 * prototype-polluting keys (`__proto__`, `constructor`, `prototype`).
 * If processing user-controlled input, sanitize the result with
 * `removePrototype` after calling this function.
 *
 * @example
 * ```typescript
 * parseQueryString("?page=1&q=search");
 * // { page: "1", q: "search" }
 *
 * parseQueryString("foo=bar&baz=qux");
 * // { foo: "bar", baz: "qux" }
 *
 * parseQueryString("https://example.com?a=1&b=2");
 * // { a: "1", b: "2" }
 * ```
 */
export const parseQueryString = (query: string): Record<string, string> => {
  let searchString = query;
  if (query.includes("://")) {
    searchString = new URL(query).search;
  }

  const parameters = new URLSearchParams(searchString);
  const result: Record<string, string> = {};
  for (const [key, value] of parameters) {
    result[key] = value;
  }
  return result;
};

File: package/umt_go/src/urlutil/parse_query_string.go
package urlutil

import (
	"net/url"
	"strings"
)

// ParseQueryString parses a query string into a key/value map.
//
// It accepts either a full URL or a raw query string (with or without a leading
// "?"), mirroring the TypeScript parseQueryString. When the input contains
// "://" it is treated as a URL and only the query portion is parsed. Parsing
// follows the application/x-www-form-urlencoded rules of URLSearchParams: pairs
// are split on "&", names and values are split on the first "=", "+" decodes to
// a space, and percent-escapes are decoded. A name without "=" maps to an empty
// value; duplicate names keep the last value, matching the Record returned by
// the TypeScript original. Bytes that cannot be percent-decoded are kept
// verbatim, like URLSearchParams.
//
// Prototype-pollution note: like the TypeScript original, this function does not
// filter dangerous keys ("__proto__", "constructor", "prototype"); sanitize the
// result before trusting user-controlled input.
//
// Example:
//
//	ParseQueryString("?page=1&q=search")          // {"page": "1", "q": "search"}
//	ParseQueryString("foo=bar&baz=qux")           // {"foo": "bar", "baz": "qux"}
//	ParseQueryString("https://example.com?a=1&b=2") // {"a": "1", "b": "2"}
func ParseQueryString(query string) map[string]string {
	searchString := query
	if strings.Contains(query, "://") {
		if parsed, err := url.Parse(query); err == nil {
			searchString = parsed.RawQuery
		} else {
			searchString = ""
		}
	}

	searchString = strings.TrimPrefix(searchString, "?")

	result := make(map[string]string)
	if searchString == "" {
		return result
	}

	for _, pair := range strings.Split(searchString, "&") {
		if pair == "" {
			continue
		}
		name := pair
		value := ""
		if equalIndex := strings.IndexByte(pair, '='); equalIndex >= 0 {
			name = pair[:equalIndex]
			value = pair[equalIndex+1:]
		}
		result[decodeFormComponent(name)] = decodeFormComponent(value)
	}

	return result
}

// decodeFormComponent decodes a single application/x-www-form-urlencoded token,
// turning "+" into a space and resolving percent-escapes. Tokens that fail to
// decode are returned with only the "+" substitution applied, matching the
// lenient behavior of URLSearchParams.
func decodeFormComponent(component string) string {
	if decoded, err := url.QueryUnescape(component); err == nil {
		return decoded
	}
	return strings.ReplaceAll(component, "+", " ")
}
